Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp

```diff
@@ -0,0 +1,248 @@
+{
+    RenderSVGHiddenContainer::styleDidChange(diff, oldStyle);
+
+    if (!m_registered) {
+        m_registered = true;
+        registerResource();
+    }
+}
+
+void RenderSVGResourceContainer::idChanged()
+{
+    // Invalidate all our current clients.
+    removeAllClientsFromCache();
+
+    // Remove old id, that is guaranteed to be present in cache.
+    treeScopeForSVGReferences().removeSVGResource(m_id);
+    m_id = element().getIdAttribute();
+
+    registerResource();
+}
+
+void RenderSVGResourceContainer::markAllClientsForRepaint()
+{
+    markAllClientsForInvalidation(RepaintInvalidation);
+}
+
+void RenderSVGResourceContainer::markAllClientsForInvalidation(InvalidationMode mode)
+{
+    // FIXME: Style invalidation should either be a pre-layout task or this function
+    // should never get called while in layout. See webkit.org/b/208903.
+    if ((m_clients.isEmptyIgnoringNullReferences() && m_clientLayers.isEmptyIgnoringNullReferences()) || m_isInvalidating)
+        return;
+
+    SetForScope<bool> isInvalidating(m_isInvalidating, true);
+
+    bool needsLayout = mode == LayoutAndBoundariesInvalidation;
+    bool markForInvalidation = mode != ParentOnlyInvalidation;
+    auto* root = SVGRenderSupport::findTreeRootObject(*this);
+
+    for (auto& client : m_clients) {
+        // We should not mark any client outside the current root for invalidation
+        if (root != SVGRenderSupport::findTreeRootObject(client))
+            continue;
+
+        if (is<RenderSVGResourceContainer>(client)) {
+            downcast<RenderSVGResourceContainer>(client).removeAllClientsFromCache(markForInvalidation);
+            continue;
+        }
+
+        if (markForInvalidation)
+            markClientForInvalidation(client, mode);
+
+        RenderSVGResource::markForLayoutAndParentResourceInvalidation(client, needsLayout);
+    }
+
+    markAllClientLayersForInvalidation();
+}
+
+void RenderSVGResourceContainer::markAllClientLayersForInvalidation()
+{
+    if (m_clientLayers.isEmptyIgnoringNullReferences())
+        return;
+
+    auto& document = (*m_clientLayers.begin()).renderer().document();
+    if (!document.view() || document.renderTreeBeingDestroyed())
+        return;
+
+    auto inLayout = document.view()->layoutContext().isInLayout();
+    for (auto& clientLayer : m_clientLayers) {
+        // FIXME: We should not get here while in layout. See webkit.org/b/208903.
+        // Repaint should also be triggered through some other means.
+        if (inLayout) {
+            clientLayer.renderer().repaint();
+            continue;
+        }
+        if (auto* enclosingElement = clientLayer.enclosingElement())
+            enclosingElement->invalidateStyleAndLayerComposition();
+        clientLayer.renderer().repaint();
+    }
+}
+
+void RenderSVGResourceContainer::markClientForInvalidation(RenderObject& client, InvalidationMode mode)
+{
+    ASSERT(!m_clients.isEmptyIgnoringNullReferences());
+
+    switch (mode) {
```

Comment:
I doubt we still need 'LayoutAndBoundaries'/'BoundariesInvalidation' - best to not start at all with this kind of code.